Failed to generate a Change Paper that included data sets - incorrect signature on Data Set class

# Description

When generating a change paper that includes data sets, an exception occurs:

```
2024-06-20 12:33:50,814 [080-exec-3] ERROR o.g.web.errors.GrailsExceptionResolver   : MissingMethodException occurred when processing request: [GET] /api/nhsdd/b0b6acf4-e690-4dfe-a1b0-22f725f53953/publish/changePaper - parameters:
dataSets: true
No signature of method: uk.nhs.digital.maurodatamapper.datadictionary.NhsDDDataSetClass.outputAsCdsHtml() is applicable for argument types: (groovy.xml.MarkupBuilder) values: [groovy.xml.MarkupBuilder@53c2e215]. Stacktrace follows:

java.lang.reflect.InvocationTargetException: null
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at org.grails.core.DefaultGrailsControllerClass$ReflectionInvoker.invoke(DefaultGrailsControllerClass.java:211)
	at org.grails.core.DefaultGrailsControllerClass.invoke(DefaultGrailsControllerClass.java:188)
	at org.grails.web.mapping.mvc.UrlMappingsInfoHandlerAdapter.handle(UrlMappingsInfoHandlerAdapter.groovy:90)
	at org.springframework.web.servlet.DispatcherServlet.doDispatch(DispatcherServlet.java:1071)
	at org.springframework.web.servlet.DispatcherServlet.doService(DispatcherServlet.java:964)
	at org.springframework.web.servlet.FrameworkServlet.processRequest(FrameworkServlet.java:1006)
	at org.springframework.web.servlet.FrameworkServlet.doGet(FrameworkServlet.java:898)
	at org.springframework.web.servlet.FrameworkServlet.service(FrameworkServlet.java:883)
	at org.grails.web.servlet.mvc.GrailsWebRequestFilter.doFilterInternal(GrailsWebRequestFilter.java:77)
	at org.grails.web.filters.HiddenHttpMethodFilter.doFilterInternal(HiddenHttpMethodFilter.java:67)
	at java.base/java.lang.Thread.run(Thread.java:840)
Caused by: groovy.lang.MissingMethodException: No signature of method: uk.nhs.digital.maurodatamapper.datadictionary.NhsDDDataSetClass.outputAsCdsHtml() is applicable for argument types: (groovy.xml.MarkupBuilder) values: [groovy.xml.MarkupBuilder@53c2e215]
	at uk.nhs.digital.maurodatamapper.datadictionary.NhsDDDataSet$_outputAsCdsHtml_closure4.doCall(NhsDDDataSet.groovy:179)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.nhs.digital.maurodatamapper.datadictionary.NhsDDDataSet.outputAsCdsHtml(NhsDDDataSet.groovy:178)
	at uk.nhs.digital.maurodatamapper.datadictionary.NhsDDDataSet.getStructureAsHtml(NhsDDDataSet.groovy:163)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.nhs.digital.maurodatamapper.datadictionary.NhsDataDictionaryComponent$Trait$Helper$_getChanges_closure7$_closure24.doCall(NhsDataDictionaryComponent.groovy:262)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at groovy.util.BuilderSupport.doInvokeMethod(BuilderSupport.java:153)
	at groovy.util.BuilderSupport.invokeMethod(BuilderSupport.java:74)
	at uk.nhs.digital.maurodatamapper.datadictionary.NhsDataDictionaryComponent$Trait$Helper$_getChanges_closure7.doCall(NhsDataDictionaryComponent.groovy:259)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at groovy.util.BuilderSupport.doInvokeMethod(BuilderSupport.java:153)
	at groovy.util.BuilderSupport.invokeMethod(BuilderSupport.java:74)
	at uk.nhs.digital.maurodatamapper.datadictionary.NhsDataDictionaryComponent$Trait$Helper.getChanges(NhsDataDictionaryComponent.groovy:258)
	at uk.nhs.digital.maurodatamapper.datadictionary.publish.changePaper.ChangePaper$_compareMaps_closure4.doCall(ChangePaper.groovy:326)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.nhs.digital.maurodatamapper.datadictionary.publish.changePaper.ChangePaper.compareMaps(ChangePaper.groovy:324)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.nhs.digital.maurodatamapper.datadictionary.publish.changePaper.ChangePaper.calculateChanges(ChangePaper.groovy:267)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at uk.nhs.digital.maurodatamapper.datadictionary.publish.changePaper.ChangePaper.<init>(ChangePaper.groovy:227)
	at java.base/jdk.internal.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at uk.nhs.digital.maurodatamapper.datadictionary.publish.changePaper.ChangePaperPdfUtility.generateChangePaper(ChangePaperPdfUtility.groovy:69)
	at uk.ac.ox.softeng.maurodatamapper.plugins.nhsdd.NhsDataDictionaryService.$tt__generateChangePaper(NhsDataDictionaryService.groovy:877)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at grails.gorm.transactions.GrailsTransactionTemplate$2.doInTransaction(GrailsTransactionTemplate.groovy:94)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at grails.gorm.transactions.GrailsTransactionTemplate.execute(GrailsTransactionTemplate.groovy:91)
	at uk.ac.ox.softeng.maurodatamapper.plugins.nhsdd.NhsDataDictionaryController.generateChangePaper(NhsDataDictionaryController.groovy:137)
	... 13 common frames omitted
```

# Steps to replicate

1. Sign in to the Orchestrator
2. Click on the "Branches" navigation link
3. Select a branch
4. Click on the "Publish" tab
5. Under "Generate change paper (with Data Set definitions)", click the "Generate" button

--- a/grails-app/services/uk/ac/ox/softeng/maurodatamapper/plugins/nhsdd/NhsDataDictionaryService.groovy
+++ b/grails-app/services/uk/ac/ox/softeng/maurodatamapper/plugins/nhsdd/NhsDataDictionaryService.groovy
@@ -867,7 +867,7 @@ class NhsDataDictionaryService {
             previousVersion = versionedFolderService.getFinalisedParent(thisDictionary)
         }
         NhsDataDictionary thisDataDictionary = buildDataDictionary(thisDictionary.id)
-        NhsDataDictionary previousDataDictionary = buildDataDictionary(previousVersion.id)
+        NhsDataDictionary previousDataDictionary = previousVersion ? buildDataDictionary(previousVersion.id) : null
 
         Path outputPath = Paths.get(getTestOutputPath())
         if(!isTest) {

--- a/src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/publish/changePaper/ChangePaper.groovy
+++ b/src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/publish/changePaper/ChangePaper.groovy
@@ -247,29 +247,29 @@ class ChangePaper {
     private List<StereotypedChange> calculateChanges(NhsDataDictionary thisDataDictionary, NhsDataDictionary previousDataDictionary, boolean includeDataSets = false) {
         List<StereotypedChange> changedItems = []
         changedItems += new StereotypedChange(stereotypeName: "NHS Business Definition",
-              changedItems: compareMaps(thisDataDictionary.businessDefinitions, previousDataDictionary.businessDefinitions))
+              changedItems: compareMaps(thisDataDictionary.businessDefinitions, previousDataDictionary?.businessDefinitions))
         changedItems += new StereotypedChange(stereotypeName: "Supporting Information",
-              changedItems: compareMaps(thisDataDictionary.supportingInformation, previousDataDictionary.supportingInformation))
+              changedItems: compareMaps(thisDataDictionary.supportingInformation, previousDataDictionary?.supportingInformation))
 
         StereotypedChange attributeChange = new StereotypedChange(stereotypeName: "Attribute",
-                changedItems: compareMaps(thisDataDictionary.attributes, previousDataDictionary.attributes))
+                changedItems: compareMaps(thisDataDictionary.attributes, previousDataDictionary?.attributes))
 
         changedItems += attributeChange
 
         StereotypedChange elementChange = new StereotypedChange(stereotypeName: "Data Element",
-                changedItems: compareMaps(thisDataDictionary.elements, previousDataDictionary.elements))
+                changedItems: compareMaps(thisDataDictionary.elements, previousDataDictionary?.elements))
 
         changedItems += elementChange
 
         changedItems += new StereotypedChange(stereotypeName: "Class",
-              changedItems: compareMaps(thisDataDictionary.classes, previousDataDictionary.classes))
+              changedItems: compareMaps(thisDataDictionary.classes, previousDataDictionary?.classes))
 
-        List<ChangedItem> dataSetChanges = compareMaps(thisDataDictionary.dataSets, previousDataDictionary.dataSets, includeDataSets)
+        List<ChangedItem> dataSetChanges = compareMaps(thisDataDictionary.dataSets, previousDataDictionary?.dataSets, includeDataSets)
 
         changedItems += new StereotypedChange(stereotypeName: "Data Set",
               changedItems: dataSetChanges)
         changedItems += new StereotypedChange(stereotypeName: "Data Set Constraint",
-              changedItems: compareMaps(thisDataDictionary.dataSetConstraints, previousDataDictionary.dataSetConstraints))
+              changedItems: compareMaps(thisDataDictionary.dataSetConstraints, previousDataDictionary?.dataSetConstraints))
 
         if(includeDataSets && dataSetChanges.size() > 0) {
             Set<NhsDDDataSet> changedDataSets = [] as Set<NhsDDDataSet>
@@ -322,7 +322,7 @@ class ChangePaper {
         boolean includeDataSets = false) {
         List<ChangedItem> changedItems = []
         newList.sort{ it.key }.each {name, component ->
-            NhsDataDictionaryComponent previousComponent = oldList[name]
+            NhsDataDictionaryComponent previousComponent = oldList?[name]
             List<Change> changes = component.getChanges(previousComponent, includeDataSets)
             if (changes) {
                 changedItems.add(new ChangedItem(

--- a/src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/publish/changePaper/ChangePaperPdfUtility.groovy
+++ b/src/main/groovy/uk/nhs/digital/maurodatamapper/datadictionary/publish/changePaper/ChangePaperPdfUtility.groovy
@@ -51,21 +51,22 @@ class ChangePaperPdfUtility {
         }
 
 
-
-        previousDataDictionary.getAllComponents().each { it ->
-            if(!pathLookup[it.getMauroPath()]) {
-                ditaProject.addExternalKey(it.getDitaKey(), it.getDataDictionaryUrl())
-                pathLookup[it.getMauroPath()] = it
+        if (previousDataDictionary) {
+            previousDataDictionary.getAllComponents().each {it ->
+                if (!pathLookup[it.getMauroPath()]) {
+                    ditaProject.addExternalKey(it.getDitaKey(), it.getDataDictionaryUrl())
+                    pathLookup[it.getMauroPath()] = it
+                }
             }
         }
-
         thisDataDictionary.getAllComponents().each {it ->
             it.replaceLinksInDefinition(pathLookup)
         }
-        previousDataDictionary.getAllComponents().each {it ->
-            it.replaceLinksInDefinition(pathLookup)
+        if (previousDataDictionary) {
+            previousDataDictionary.getAllComponents().each {it ->
+                it.replaceLinksInDefinition(pathLookup)
+            }
         }
-
         ChangePaper changePaper = new ChangePaper(thisDataDictionary, previousDataDictionary, includeDataSets)
 
 

